Pinta.Effects/Effects/ForwardErrorDiffusionDitheringEffect.cs

```diff
@@ -0,0 +1,325 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Cairo;
+using Pinta.Core;
+using Pinta.Gui.Widgets;
+
+namespace Pinta.Effects;
+
+public sealed class ForwardErrorDiffusionDitheringEffect : BaseEffect
+{
+	public override string Name => Translations.GetString ("Dithering");
+	public override bool IsConfigurable => true;
+	// TODO: Icon
+	public override string EffectMenuCategory => Translations.GetString ("Test"); // TODO:
+	public ForwardErrorDiffusionDitheringData Data => (ForwardErrorDiffusionDitheringData) EffectData!; // NRT - Set in constructor
+
+	public ForwardErrorDiffusionDitheringEffect ()
+	{
+		EffectData = new ForwardErrorDiffusionDitheringData ();
+	}
+
+	public override void Render (ImageSurface src, ImageSurface dest, ReadOnlySpan<RectangleI> rois)
+	{
+		var src_data = src.GetReadOnlyPixelData ();
+		var src_copy = new ColorBgra[src_data.Length];
+		src_data.CopyTo (src_copy);
+		var dst_data = dest.GetPixelData ();
+		foreach (var rect in rois) {
+			for (int y = 0; y < rect.Height - Data.DiffusionMatrix.RowsBelow; y++) {
+				for (int x = Data.DiffusionMatrix.ColumnsToLeft; x < rect.Width - Data.DiffusionMatrix.ColumnsToRight; x++) {
```

Comment:
I'm not really sure about the indexing here, this seems to ignore the left and top coordinates of the rectangle?
Also I think the dithering should likely still be applied to the first pixel(s) of the row?